src/processor/operator/recursive_extend/variable_length_recursive_join.cpp

```diff
@@ -0,0 +1,71 @@
+#include "processor/operator/recursive_extend/variable_length_recursive_join.h"
+
+namespace kuzu {
+namespace processor {
+
+void VariableLengthRecursiveJoin::initLocalStateInternal(
+    ResultSet* resultSet_, ExecutionContext* context) {
+    BaseRecursiveJoin::initLocalStateInternal(resultSet_, context);
+    auto maxNodeOffset = nodeTable->getMaxNodeOffset(transaction);
+    bfsMorsel = std::make_unique<VariableLengthBFSMorsel>(
+        maxNodeOffset, lowerBound, upperBound, sharedState->semiMask.get());
+    bfsMorsel->resetState();
+}
+
+bool VariableLengthRecursiveJoin::scanOutput() {
+    auto morsel = (VariableLengthBFSMorsel*)bfsMorsel.get();
+    if (outputCursor == morsel->dstNodeOffsets.size()) {
+        return false;
+    }
+    auto vectorSize = 0u;
+    while (vectorSize != common::DEFAULT_VECTOR_CAPACITY &&
+           outputCursor < morsel->dstNodeOffsets.size()) {
+        auto offset = morsel->dstNodeOffsets[outputCursor];
+        auto& numPath = morsel->dstNodeOffset2NumPath.at(offset);
+        auto numPathFitInCurrentVector =
+            std::min<uint64_t>(common::DEFAULT_VECTOR_CAPACITY - vectorSize, numPath);
+        for (auto i = 0u; i < numPathFitInCurrentVector; ++i) {
+            dstNodeIDVector->setValue<common::nodeID_t>(
+                vectorSize, common::nodeID_t{offset, nodeTable->getTableID()});
+            vectorSize++;
+        }
+        numPath -= numPathFitInCurrentVector;
+        outputCursor += numPath == 0; // No more path to scan under current cursor
+    }
+    dstNodeIDVector->state->initOriginalAndSelectedSize(vectorSize);
+    return true;
+}
+
+void VariableLengthRecursiveJoin::computeBFS(ExecutionContext* context) {
```

Comment:
You could really write 1 function for computeBFS. It's almost the same code across VariableLengthRecursiveJoin and ShortestPathRecursiveJoin. The code would look like:
```
void BaseRecursiveJoin::computeBFS(ExecutionContext* context) {
    auto morsel = bfsMorsel.get();
    auto nodeID = srcNodeIDVector->getValue<common::nodeID_t>(
        srcNodeIDVector->state->selVector->selectedPositions[0]);
    morsel->markSrc(nodeID.offset);
    while (!morsel->isComplete()) {
        auto nodeOffset = morsel->getNextNodeOffset();
        if (nodeOffset != common::INVALID_NODE_OFFSET) {
            // Found a starting node from current frontier.
            scanBFSLevel->setNodeID(common::nodeID_t{nodeOffset, nodeTable->getTableID()});
            while (recursiveRoot->getNextTuple(context)) { // Exhaust recursive plan.
                updateVisitedNode(morsel, nodeOffset);
            }
        } else {
            // Otherwise move to the next frontier.
            morsel->finishBFSMorsel();
        }
    }
}
```
The only difference is that VariableLengthRecursiveJoin does some work related to multiplicity. One happens in the updateVisitedNode (or updateVisitedState in ShortestPathRecursiveJoin). You can pass in nodeOffset to a common virtual function updateVisitedNode and VariableLengthRecursiveJoin::updateVisitedNode can read the multiplicity there instead of outside the while loop.

The other difference is that VariableLengthRecursiveJoin has to call the `updateNumPathFromCurrentFrontier` function. You can call a virtual morsel->finishBFSMorsel(). In ShortestPathRecursiveJoin this can just be        `morsel->moveNextLevelAsCurrentLevel();` and in VariableLengthRecursiveJoin it can be two lines `morsel->moveNextLevelAsCurrentLevel();  morsel->updateNumPathFromCurrentFrontierIfNecessary();`. 

I think it's probably worth it, at least for now, to factor out the common algorithm across these functions. We can diverge later once we add different versions of these recursive joins.